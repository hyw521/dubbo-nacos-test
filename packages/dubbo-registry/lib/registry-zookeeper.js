'use strict'
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod }
  }
Object.defineProperty(exports, '__esModule', { value: true })
exports.Zk = exports.ZookeeperRegistry = void 0
const debug_1 = __importDefault(require('debug'))
const zookeeper_1 = __importDefault(require('zookeeper'))
const registry_base_1 = __importDefault(require('./registry-base'))
const timeout_1 = __importDefault(require('./timeout'))
const DUBBO_ZK_ROOT_PATH = '/dubbo'
const dlog = (0, debug_1.default)('dubbo:zookeeper~')
class ZookeeperRegistry extends registry_base_1.default {
  constructor(props) {
    super()
    dlog(`init zookeeper with %O`, props)
    ZookeeperRegistry.checkProps(props)
    this.props = props
    this.props.zkRootPath = this.props.zkRootPath || DUBBO_ZK_ROOT_PATH
    // init ready promise
    this.readyPromise = new Promise((resolve, reject) => {
      this.resolve = resolve
      this.reject = reject
    })
    this.timeout = new timeout_1.default({
      maxTimeout: this.props.timeout || 40 * 1000,
      onTimeout: () => {
        this.reject(
          new Error(`zookeeper connect ${this.props.connect} timeout`)
        )
        this.close()
      }
    })
    this.init()
  }
  // ~~~~~~~~~~~~~~~~ private ~~~~~~~~~~~~~~~~~~~~~~~~~~
  static checkProps(props) {
    if (!props.connect) {
      throw new Error(`Please specify zookeeper connect url`)
    }
  }
  init() {
    if (this.client) {
      return
    }
    // set default props value
    this.props.timeout = this.props.timeout || 40 * 1000
    this.props.debug_level =
      this.props.debug_level || zookeeper_1.default.constants.ZOO_LOG_LEVEL_WARN
    this.props.host_order_deterministic =
      this.props.host_order_deterministic || false
    dlog('connecting zookeeper with %O', this.props)
    this.client = new zookeeper_1.default(this.props)
    this.client.on('connect', async () => {
      dlog('connected with zookeeper with %s', this.props.connect)
      this.timeout.clearTimeout()
      try {
        // create root node
        await this.mkdirp(this.props.zkRootPath)
        // trigger ready promise
        this.resolve()
      } catch (err) {
        this.reject(err)
      }
    })
    this.client.on('close', () => {
      dlog(`zookeeper closed`)
      this.emitErr(new Error(`Zookeeper was closed`))
      this.close()
      this.init()
    })
    this.client.on('error', (err) => {
      dlog(`zookeeper error %s`, err)
      this.reject(err)
      this.emitErr(err)
    })
    process.nextTick(() => {
      this.client.init({})
    })
  }
  async createNode(cfg) {
    dlog(`create zookeeper node %j`, cfg)
    const { path, data = '', isPersistent = false } = cfg
    try {
      await this.client.exists(path, false)
      dlog(`${path} node was existed ~`)
    } catch (err) {
      await this.client.create(
        path,
        data,
        isPersistent
          ? zookeeper_1.default.constants.ZOO_PERSISTENT
          : zookeeper_1.default.constants.ZOO_EPHEMERAL
      )
    }
  }
  async mkdirp(path) {
    return new Promise((resolve, reject) => {
      this.client.mkdirp(path, (err) => {
        if (err) {
          dlog(`mkdir %s error %s`, path, err)
          reject(err)
        } else {
          dlog('mkdir %s ok', path)
          resolve()
        }
      })
    })
  }
  wrapWatch(dubboInterface) {
    const servicePath = `${this.props.zkRootPath}/${dubboInterface}/providers`
    return async (type, state) => {
      dlog('wrapWatch %s %d %d', servicePath, type, state)
      await this.findDubboServiceUrl(dubboInterface)
      this.emitData(this.dubboServiceUrlMap)
    }
  }
  // ~~~~~~~~~~~~~~~~ public ~~~~~~~~~~~~~~~~~~~~~~~~~~
  getProps() {
    return this.props
  }
  ready() {
    return this.readyPromise
  }
  async findDubboServiceUrls(dubboInterfaces) {
    dlog('find dubbo service urls => %O', dubboInterfaces)
    await Promise.all(
      dubboInterfaces.map((dubboInterface) =>
        this.findDubboServiceUrl(dubboInterface)
      )
    )
    this.emitData(this.dubboServiceUrlMap)
  }
  async findDubboServiceUrl(dubboInterface) {
    const servicePath = `${this.props.zkRootPath}/${dubboInterface}/providers`
    const urls = (
      await this.client
        .w_get_children(servicePath, this.wrapWatch(dubboInterface))
        .catch((err) => {
          dlog(
            `get beehive service urls err %s %s %s`,
            servicePath,
            dubboInterface,
            err
          )
          return []
        })
    )
      .map((v) => decodeURIComponent(v))
      .filter((v) => v.startsWith('dubbo://'))
    this.dubboServiceUrlMap.set(dubboInterface, urls)
  }
  async registerServices(services) {
    for (let { dubboServiceInterface, dubboServiceUrl } of services) {
      // create service root path
      const serviceRootPath = `${this.props.zkRootPath}/${dubboServiceInterface}/providers`
      await this.mkdirp(serviceRootPath)
      // create service node
      await this.createNode({
        path: `${serviceRootPath}/${encodeURIComponent(dubboServiceUrl)}`
      })
    }
  }
  async registerConsumers(consumers) {
    dlog('registry consumers => %O', consumers)
    const dubboInterfaces = new Set()
    // registry consumer
    for (let { dubboServiceInterface, dubboServiceUrl } of consumers) {
      dubboInterfaces.add(dubboServiceInterface)
      // create consumer root path
      const consumerRootPath = `${this.props.zkRootPath}/${dubboServiceInterface}/consumers`
      await this.mkdirp(consumerRootPath)
      // create service node
      await this.createNode({
        path: `${consumerRootPath}/${encodeURIComponent(dubboServiceUrl)}`
      })
    }
    await this.findDubboServiceUrls([...dubboInterfaces])
  }
  close() {
    var _a, _b
    this.timeout.clearTimeout()
    ;(_a = this.client) === null || _a === void 0
      ? void 0
      : _a.removeAllListeners()
    ;(_b = this.client) === null || _b === void 0 ? void 0 : _b.close()
    this.client = null
  }
  getClient() {
    return this.client
  }
}
exports.ZookeeperRegistry = ZookeeperRegistry
function Zk(props) {
  return new ZookeeperRegistry(props)
}
exports.Zk = Zk
