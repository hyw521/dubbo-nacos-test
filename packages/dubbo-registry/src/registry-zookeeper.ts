/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import debug from 'debug'
import Zookeeper from 'zookeeper'
import { IRegistry } from './registry'
import BaseRegistry from './registry-base'
import Timeout from './timeout'
import {
  INodeProps,
  IZkClientConfig,
  TDubboInterface,
  TDubboUrl
} from './types'

const DUBBO_ZK_ROOT_PATH: string = '/dubbo'
const dlog = debug('dubbo:zookeeper~')

export class ZookeeperRegistry
  extends BaseRegistry
  implements IRegistry<Zookeeper>
{
  private readonly props: IZkClientConfig
  private client: Zookeeper
  private timeout: Timeout

  private readonly readyPromise: Promise<void>
  private resolve: Function
  private reject: Function

  constructor(props: IZkClientConfig) {
    super()
    dlog(`init zookeeper with %O`, props)
    ZookeeperRegistry.checkProps(props)
    this.props = props

    this.props.zkRootPath = this.props.zkRootPath || DUBBO_ZK_ROOT_PATH

    // init ready promise
    this.readyPromise = new Promise((resolve, reject) => {
      this.resolve = resolve
      this.reject = reject
    })

    this.timeout = new Timeout({
      maxTimeout: this.props.timeout || 40 * 1000,
      onTimeout: () => {
        this.reject(
          new Error(`zookeeper connect ${this.props.connect} timeout`)
        )
        this.close()
      }
    })
    this.init()
  }

  // ~~~~~~~~~~~~~~~~ private ~~~~~~~~~~~~~~~~~~~~~~~~~~
  private static checkProps(props: IZkClientConfig) {
    if (!props.connect) {
      throw new Error(`Please specify zookeeper connect url`)
    }
  }

  private init() {
    if (this.client) {
      return
    }

    // set default props value
    this.props.timeout = this.props.timeout || 40 * 1000
    this.props.debug_level =
      this.props.debug_level || Zookeeper.constants.ZOO_LOG_LEVEL_WARN
    this.props.host_order_deterministic =
      this.props.host_order_deterministic || false
    dlog('connecting zookeeper with %O', this.props)

    this.client = new Zookeeper(this.props)

    this.client.on('connect', async () => {
      dlog('connected with zookeeper with %s', this.props.connect)
      this.timeout.clearTimeout()

      try {
        // create root node
        await this.mkdirp(this.props.zkRootPath)
        // trigger ready promise
        this.resolve()
      } catch (err) {
        this.reject(err)
      }
    })

    this.client.on('close', () => {
      dlog(`zookeeper closed`)
      this.emitErr(new Error(`Zookeeper was closed`))
      this.close()
      this.init()
    })

    this.client.on('error', (err) => {
      dlog(`zookeeper error %s`, err)
      this.reject(err)
      this.emitErr(err)
    })

    process.nextTick(() => {
      this.client.init({})
    })
  }

  private async createNode(cfg: INodeProps) {
    dlog(`create zookeeper node %j`, cfg)
    const { path, data = '', isPersistent = false } = cfg
    try {
      await this.client.exists(path, false)
      dlog(`${path} node was existed ~`)
    } catch (err) {
      await this.client.create(
        path,
        data,
        isPersistent
          ? Zookeeper.constants.ZOO_PERSISTENT
          : Zookeeper.constants.ZOO_EPHEMERAL
      )
    }
  }

  private async mkdirp(path: string): Promise<void> {
    return new Promise((resolve, reject) => {
      this.client.mkdirp(path, (err) => {
        if (err) {
          dlog(`mkdir %s error %s`, path, err)
          reject(err)
        } else {
          dlog('mkdir %s ok', path)
          resolve()
        }
      })
    })
  }

  private wrapWatch(dubboInterface: string) {
    const servicePath = `${this.props.zkRootPath}/${dubboInterface}/providers`
    return async (type: number, state: number) => {
      dlog('wrapWatch %s %d %d', servicePath, type, state)
      await this.findDubboServiceUrl(dubboInterface)
      this.emitData(this.dubboServiceUrlMap)
    }
  }

  // ~~~~~~~~~~~~~~~~ public ~~~~~~~~~~~~~~~~~~~~~~~~~~

  getProps() {
    return this.props
  }

  ready(): Promise<void> {
    return this.readyPromise
  }

  async findDubboServiceUrls(dubboInterfaces: Array<string>) {
    dlog('find dubbo service urls => %O', dubboInterfaces)
    await Promise.all(
      dubboInterfaces.map((dubboInterface) =>
        this.findDubboServiceUrl(dubboInterface)
      )
    )
    this.emitData(this.dubboServiceUrlMap)
  }

  async findDubboServiceUrl(dubboInterface: string) {
    const servicePath = `${this.props.zkRootPath}/${dubboInterface}/providers`
    const urls = (
      await this.client
        .w_get_children(servicePath, this.wrapWatch(dubboInterface))
        .catch((err) => {
          dlog(
            `get beehive service urls err %s %s %s`,
            servicePath,
            dubboInterface,
            err
          )
          return []
        })
    )
      .map((v: string) => decodeURIComponent(v))
      .filter((v: string) => v.startsWith('dubbo://'))
    this.dubboServiceUrlMap.set(dubboInterface, urls)
  }

  async registerServices(
    services: Array<{
      dubboServiceInterface: TDubboInterface
      dubboServiceUrl: TDubboUrl
    }>
  ) {
    for (let { dubboServiceInterface, dubboServiceUrl } of services) {
      // create service root path
      const serviceRootPath = `${this.props.zkRootPath}/${dubboServiceInterface}/providers`
      await this.mkdirp(serviceRootPath)
      // create service node
      await this.createNode({
        path: `${serviceRootPath}/${encodeURIComponent(dubboServiceUrl)}`
      })
    }
  }

  async registerConsumers(
    consumers: Array<{
      dubboServiceInterface: TDubboInterface
      dubboServiceUrl: TDubboUrl
    }>
  ) {
    dlog('registry consumers => %O', consumers)
    const dubboInterfaces = new Set<string>()
    // registry consumer
    for (let { dubboServiceInterface, dubboServiceUrl } of consumers) {
      dubboInterfaces.add(dubboServiceInterface)
      // create consumer root path
      const consumerRootPath = `${this.props.zkRootPath}/${dubboServiceInterface}/consumers`
      await this.mkdirp(consumerRootPath)
      // create service node
      await this.createNode({
        path: `${consumerRootPath}/${encodeURIComponent(dubboServiceUrl)}`
      })
    }

    await this.findDubboServiceUrls([...dubboInterfaces])
  }

  close(): void {
    this.timeout.clearTimeout()
    this.client?.removeAllListeners()
    this.client?.close()
    this.client = null
  }

  getClient() {
    return this.client
  }
}

export function Zk(props: IZkClientConfig) {
  return new ZookeeperRegistry(props)
}
