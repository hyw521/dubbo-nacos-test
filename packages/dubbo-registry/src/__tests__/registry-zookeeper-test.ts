/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Zookeeper from 'zookeeper'
import { Zk } from '../registry-zookeeper'

describe('test zookeeper registry', () => {
  it('test throw connection url error', () => {
    try {
      Zk({
        connect: ''
      })
    } catch (err) {
      expect(err.message).toEqual('Please specify zookeeper connect url')
    }
  })

  it('test zk props and ready ok', async () => {
    const zk = Zk({
      connect: 'localhost:2181'
    })
    expect(zk.getProps()).toEqual({
      connect: 'localhost:2181',
      timeout: 40000,
      debug_level: Zookeeper.constants.ZOO_LOG_LEVEL_WARN,
      host_order_deterministic: false,
      zkRootPath: '/dubbo'
    })
    const res = await zk.ready()
    expect(res).toBeUndefined()

    zk.close()
  })

  it('test zk with zkRootPath and ready ok', async () => {
    const zk = Zk({
      connect: 'localhost:2181',
      zkRootPath: '/test/com.demo.dubbo'
    })

    expect(zk.getProps()).toEqual({
      connect: 'localhost:2181',
      timeout: 40000,
      debug_level: Zookeeper.constants.ZOO_LOG_LEVEL_WARN,
      host_order_deterministic: false,
      zkRootPath: '/test/com.demo.dubbo'
    })
    const res = await zk.ready()
    expect(res).toBeUndefined()

    zk.close()
  })

  it('test zk ready failed', async () => {
    const zk = Zk({
      connect: 'localhost:2181'
    })
    zk.subscribe({
      onData() {},
      onError(err) {
        expect(err.message).toEqual(`zk could not connect`)
      }
    })
    const client = zk.getClient()
    //@ts-ignore
    client.mockConnectErr()

    const err = await zk.ready().catch((err) => err)
    expect(err instanceof Error).toBeTruthy()

    zk.close()
  })

  it('test registerServices', async () => {
    const services = [
      {
        dubboServiceInterface: 'org.apache.demo.service.HelloService',
        dubboServiceUrl:
          'dubbo://127.0.01:20880/org.apache.demo.HelloService?interface=org.apache.demo.service.HelloService&methods=sayHello,test,echo,getUserInfo&version=1.0.0'
      },
      {
        dubboServiceInterface: 'org.apache.demo.service.UserService',
        dubboServiceUrl:
          'dubbo://127.0.01:20880/org.apache.demo.UserService?interface=org.apache.demo.service.UserService&methods=sayHello,test,echo,getUserInfo&version=1.0.0'
      }
    ]

    const zk = Zk({
      connect: 'localhost:2181'
    })

    await zk.ready()
    await zk.registerServices(services)

    zk.close()
  })

  it('test registryConsumer', async () => {
    const services = [
      {
        dubboServiceInterface: 'org.apache.demo.service.HelloService',
        dubboServiceUrl:
          'consumer://127.0.01:20880/org.apache.demo.HelloService?version=1.0.0'
      },
      {
        dubboServiceInterface: 'org.apache.demo.service.UserService',
        dubboServiceUrl:
          'consumer://127.0.01:20880/org.apache.demo.UserService?version=1.0.0'
      }
    ]
    const zk = Zk({ connect: 'localhost:2181' })
    zk.subscribe({
      onData(data) {
        expect(data).toMatchSnapshot()
      },
      onError(err) {
        expect(err).toMatchSnapshot()
      }
    })
    await zk.ready()
    await zk.registerConsumers(services)
    zk.close()
  })
})
