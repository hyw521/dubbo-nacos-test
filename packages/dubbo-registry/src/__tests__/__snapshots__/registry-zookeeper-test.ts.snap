// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test zookeeper registry test registryConsumer 1`] = `
Map {
  "org.apache.demo.service.HelloService" => Array [],
  "org.apache.demo.service.UserService" => Array [],
}
`;
